Sprint №9 HomeWork

Vozzhd opened the pull request.

Vozzhd committed:
HomeWork #1 is done.
Added styles, changed themes and icon colors from "@color/***" to "?attr/"

Vozzhd committed:
Sprint №9 HomeWork №2 is done.
Added:
-listeners to buttons in SettingsActivity, all buttons in this activity works
-string-resourses
-
Changed:
-corrected theme in AndroidManifest.xml

Vozzhd committed:
Sprint №9 HomeWork №3 is done.
Added:
-icons
-custom background of TextView
-string resourses
Changed:
-add work with edit text in settingsActivity
-add new colors in themes Tertiary(day-gray/night-black)

Vozzhd committed:
Sprint №9 HomeWork №4 is done.
Added:
-save userInput in SearchActivity

Vozzhd committed:
Sprint №9 HomeWork №4 is done.UPDATED
Added:
-parent for bigButton style in activity_main
-hideKeyboard fun, after "clearButton" pressed
Changed:
-move restore of savedInput from onCreate to onRestoreInstanceState

vlad-zotov commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,11 +1,77 @@
+
+        }
+
+        fun clearButtonVisibility(textInView: CharSequence?): Int {
```
Привет! Советую называть функции так, чтобы было понятно какую логику они выполняют. Эта функция возвращает значение для видимости кнопки. Поэтому более правильным было бы название `getClearButtonVisibility`. Логичные и полные названия методов позволят сделать код более читаемым, это очень важно в командной разработке.

Vozzhd replied:
Ок, исправил

vlad-zotov commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,11 +1,77 @@
+        inputEditText.setText(savedInputInFindView)
+    }
+
+    private fun View.hideKeyboard() {
```
удобно выносить такие функции-расширения в отдельные файлы-утилиты. Так ты сможешь переиспользовать одни и те же функции в разных местах проекта.

Vozzhd replied:
Сделал, вынес, спасибо за совет)

vlad-zotov commented on app/src/main/java/com/practicum/playlistmaker/SettingsActivity.kt:
```diff
@@ -1,27 +1,56 @@
+            val sharedLink = getString(R.string.linkInsideShareButton)
+            intent.action = Intent.ACTION_SEND
+            intent.putExtra(Intent.EXTRA_TEXT, sharedLink);
+            intent.type = "application/octet-stream"
```
тип данных `application/octet-stream` подходит для двоичных данных. Ты передаешь строку с ссылкой для обмена, для этого тебе потребуется тип `text/plain`

Vozzhd replied:
Спасибо, скорректировал

vlad-zotov commented on app/src/main/res/drawable/custom_edit_text_rounded.xml:
```diff
@@ -0,0 +1,21 @@
+        <shape android:shape="rectangle">
+            <solid android:color="?attr/colorSecondaryContainer" />
+            <corners android:radius="8dp" />
+            " />
```
лишние символы, которые не используются в этом компоненте. Нужно удалить.

Vozzhd replied:
Эти символы оставил - используются (отвечают за цвета когда элемент НЕ в фокусе)
Удалил блок выше, который отвечал за цвета когда элемент В фокусе.

vlad-zotov commented on app/src/main/res/drawable/custom_edit_text_rounded.xml:
```diff
@@ -0,0 +1,21 @@
+    <item android:state_enabled="true" android:state_focused="true">
+        <shape android:shape="rectangle">
+            <solid android:color="?attr/colorSecondaryContainer" />
+            <corners android:radius="8dp"
```
значения размеров, отступов и скруглений необходимо помещать в ресурсы `dimens.xml`

Vozzhd replied:
Изначально речь в спринтах шла о том, что **повторяемые** значения нужно туда выносить. Значение 8dp я **нигде** больше не использую. 
Но ок, перенёс в строковый ресурс `radiusOfEditTextViewCorners`

vlad-zotov commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -1,8 +1,39 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/defaultMargin"
+        android:layout_marginTop="14dp"
```
значения размеров, отступов и скруглений необходимо помещать в ресурсы dimens.xml

Vozzhd replied:
Изначально речь в спринтах шла о том, что **повторяемые** значения нужно туда выносить. Значение 14dp я **нигде** больше не использую. 
Отступил от дизайна, сделав значение равным `defaultMargin` (16dp) . 
Визуально малозаметно, зато стало кратным 8 и теперь используется размер из ресурсов.

vlad-zotov commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -1,8 +1,39 @@
+
+    <FrameLayout
+        android:layout_width="match_parent"
+        android:layout_height="56dp"
```
значения размеров, отступов и скруглений необходимо помещать в ресурсы dimens.xml

Vozzhd replied:
Изначально речь в спринтах шла о том, что **повторяемые** значения нужно туда выносить. Значение 56dp я **нигде** больше не использую. 
Изменил значение на 60dp, и применил значение из ресурсов, такую же высоту использую для кнопок в  (defaultFrameLayoutLineHeight)

vlad-zotov commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -1,8 +1,39 @@
+        android:layout_width="match_parent"
+        android:layout_height="56dp"
+        android:layout_marginHorizontal="@dimen/defaultMargin"
+        android:layout_marginTop="64dp">
```
значения размеров, отступов и скруглений необходимо помещать в ресурсы dimens.xml

Vozzhd replied:
Изначально речь в спринтах шла о том, что **повторяемые** значения нужно туда выносить. Значение 64dp я **нигде** больше не использую. 
Но ок, вынес в строковый ресурс `marginFromTopForSearchField`

vlad-zotov commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -1,8 +1,39 @@
+
+        <ImageView
+            android:id="@+id/clearButton"
+            android:layout_width="16dp"
```
значения размеров, отступов и скруглений необходимо помещать в ресурсы dimens.xml

Vozzhd replied:
Вынес в `defaultSmallIconSize`

vlad-zotov commented on app/src/main/res/values/styles.xml:
```diff
@@ -2,13 +2,15 @@
 <resources>
 
-    <style name="bigButton">
+    <style name="bigButton" parent="Widget.MaterialComponents.Button.UnelevatedButton">
```
стили принято называть с большой буквы

Vozzhd replied:
Понял, исправил названия всех стилей.

vlad-zotov commented on app/src/main/java/com/practicum/playlistmaker/SettingsActivity.kt:
```diff
@@ -1,27 +1,56 @@
+            intent.action = Intent.ACTION_SEND
+            intent.putExtra(Intent.EXTRA_TEXT, sharedLink);
+            intent.type = "application/octet-stream"
+            startActivity(Intent.createChooser(intent, "share to:"))
```
`"share to:"` необходимо вынести в ресурсы

Vozzhd replied:
Исправил строку на "share by:", вынес в ресурс `hintInShareWindow`

Vozzhd committed:
Sprint №9 HomeWork №4 UPD after review.

Added:
-new dimension resources
-new string resources
-file with Utilities

Changed:
-renamed styles (UpperCase first Letters)
-deleted onFocus color settings in custom_edit_text_rounded.xml
-moved fun hideKeyboard from SearchActivity.kt to Utilities.kt
-rewired Id's from <ImageView> to <FrameLayout> in activity_settings.xml/SettingsActivity.kt
-changed switchCompat position in activity_settings.xml

Vozzhd commented:
Касательно критичных моментов:
1) `На экране настроек работают клики только на иконках кнопок. Для хорошего пользовательского UX необходимо обрабатывать клики на тексте кнопки.` К сожалению - в ТЗ чётко обозначен этот момент не был. Исправлено
2) `Отступ кнопки Share не соответствует дизайну` - соответствует самому первому дизайну, забыл исправить на актуальный. Исправлено
3) `Не все строки и размеры вынесены в ресурсы.` Могу вынести вообще все, но о вынесении всех размером в ресурсы речь в ТЗ не шла. Исправлено

vlad-zotov commented:
> К сожалению - в ТЗ чётко обозначен этот момент не был. Исправлено

К сожалению, косяк нашего ТЗ, внесем правки. Тут стоит держать в голове, что при разработке учебного проекта мы хотим сделать его максимально похожим по юзабилити на обычные приложения из маркета.

> Могу вынести вообще все, но о вынесении всех размером в ресурсы речь в ТЗ не шла. Исправлено

Это история больше не про ТЗ, а про удобство разработки. Во многих проектах практикуется использование устоявшегося набора размеров и отступов для UI элементов. Это позволяет лучше контролировать изменения в дизайне проекта. 

Хорошая работа! Извиняюсь за задержку с ревью.
